{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_AMR_MOB=\"/lomi_home/gaoyang/software/CompRanking/test/CompRanking/CompRanking_result/CompRanking_ERR1191817.contigs_AMR_MOB_prediction.tsv\"\n",
    "input_VF_PATH=\"/lomi_home/gaoyang/software/CompRanking/test/CompRanking/CompRanking_result/CompRanking_ERR1191817.contigs_Virulence_Pathogenic_prediction.tsv\"\n",
    "df_AMR_MOB=pd.read_csv(input_AMR_MOB,sep=\"\\t\",header=0)\n",
    "df_VF_PATH=pd.read_csv(input_VF_PATH,sep=\"\\t\",header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Virulence_factor</th>\n",
       "      <th>VF_Name</th>\n",
       "      <th>VF_Host_Bacteria</th>\n",
       "      <th>VFcategory</th>\n",
       "      <th>ESKAPE</th>\n",
       "      <th>Pathogenicity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262013</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262014</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262031</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>wbtL</td>\n",
       "      <td>LPS</td>\n",
       "      <td>Francisella tularensis</td>\n",
       "      <td>Immune modulation</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262039</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262040</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>243290 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Contig       ORF Virulence_factor VF_Name  \\\n",
       "0            k119_7       1_1                -       -   \n",
       "1            k119_8       2_1                -       -   \n",
       "2           k119_12       3_1                -       -   \n",
       "3           k119_12       3_2                -       -   \n",
       "4           k119_14       4_1                -       -   \n",
       "...             ...       ...              ...     ...   \n",
       "262013  k119_301780  123145_2                -       -   \n",
       "262014  k119_301781  123146_1                -       -   \n",
       "262031  k119_301781  123146_1             wbtL     LPS   \n",
       "262039  k119_301781  123146_2                -       -   \n",
       "262040  k119_301781  123146_3                -       -   \n",
       "\n",
       "              VF_Host_Bacteria         VFcategory    ESKAPE Pathogenicity  \n",
       "0                            -                  -         -             -  \n",
       "1                            -                  -         -    Pathogenic  \n",
       "2                            -                  -         -    Pathogenic  \n",
       "3                            -                  -         -    Pathogenic  \n",
       "4                            -                  -         -             -  \n",
       "...                        ...                ...       ...           ...  \n",
       "262013                       -                  -         -             -  \n",
       "262014                       -                  -         -    Pathogenic  \n",
       "262031  Francisella tularensis  Immune modulation  Negative    Pathogenic  \n",
       "262039                       -                  -         -    Pathogenic  \n",
       "262040                       -                  -         -    Pathogenic  \n",
       "\n",
       "[243290 rows x 8 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_VF_PATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>ORF_ID</th>\n",
       "      <th>ARG_prediction</th>\n",
       "      <th>ARG_class</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Database</th>\n",
       "      <th>ARG_rank</th>\n",
       "      <th>CompRanking_MGE_prediction</th>\n",
       "      <th>mobileOG_ID</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Major_Category</th>\n",
       "      <th>MGE_Database</th>\n",
       "      <th>coverage</th>\n",
       "      <th>MOB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240330</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240331</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240332</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240333</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240334</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>240335 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Contig    ORF_ID ARG_prediction ARG_class SNPs Database ARG_rank  \\\n",
       "0            k119_7       1_1              -         -    -        -        -   \n",
       "1            k119_8       2_1              -         -    -        -        -   \n",
       "2           k119_12       3_1              -         -    -        -        -   \n",
       "3           k119_12       3_2              -         -    -        -        -   \n",
       "4           k119_14       4_1              -         -    -        -        -   \n",
       "...             ...       ...            ...       ...  ...      ...      ...   \n",
       "240330  k119_301780  123145_1              -         -    -        -        -   \n",
       "240331  k119_301780  123145_2              -         -    -        -        -   \n",
       "240332  k119_301781  123146_1              -         -    -        -        -   \n",
       "240333  k119_301781  123146_2              -         -    -        -        -   \n",
       "240334  k119_301781  123146_3              -         -    -        -        -   \n",
       "\n",
       "       CompRanking_MGE_prediction mobileOG_ID Gene_Name Taxonomy  \\\n",
       "0                           phage           -         -        -   \n",
       "1                    unclassified           -         -        -   \n",
       "2                    unclassified           -         -        -   \n",
       "3                    unclassified           -         -        -   \n",
       "4                         plasmid           -         -        -   \n",
       "...                           ...         ...       ...      ...   \n",
       "240330                 chromosome           -         -        -   \n",
       "240331                 chromosome           -         -        -   \n",
       "240332               unclassified           -         -        -   \n",
       "240333               unclassified           -         -        -   \n",
       "240334               unclassified           -         -        -   \n",
       "\n",
       "       Major_Category MGE_Database coverage MOB  \n",
       "0                   -            -        -   -  \n",
       "1                   -            -        -   -  \n",
       "2                   -            -        -   -  \n",
       "3                   -            -        -   -  \n",
       "4                   -            -        -   -  \n",
       "...               ...          ...      ...  ..  \n",
       "240330              -            -        -   -  \n",
       "240331              -            -        -   -  \n",
       "240332              -            -        -   -  \n",
       "240333              -            -        -   -  \n",
       "240334              -            -        -   -  \n",
       "\n",
       "[240335 rows x 15 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_AMR_MOB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig_x</th>\n",
       "      <th>ORF_ID</th>\n",
       "      <th>ARG_prediction</th>\n",
       "      <th>ARG_class</th>\n",
       "      <th>SNPs</th>\n",
       "      <th>Database</th>\n",
       "      <th>ARG_rank</th>\n",
       "      <th>CompRanking_MGE_prediction</th>\n",
       "      <th>mobileOG_ID</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>...</th>\n",
       "      <th>coverage</th>\n",
       "      <th>MOB</th>\n",
       "      <th>Contig_y</th>\n",
       "      <th>ORF</th>\n",
       "      <th>Virulence_factor</th>\n",
       "      <th>VF_Name</th>\n",
       "      <th>VF_Host_Bacteria</th>\n",
       "      <th>VFcategory</th>\n",
       "      <th>ESKAPE</th>\n",
       "      <th>Pathogenicity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262036</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262037</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262038</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262039</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262040</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>262041 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Contig_x    ORF_ID ARG_prediction ARG_class SNPs Database ARG_rank  \\\n",
       "0            k119_7       1_1              -         -    -        -        -   \n",
       "1            k119_8       2_1              -         -    -        -        -   \n",
       "2           k119_12       3_1              -         -    -        -        -   \n",
       "3           k119_12       3_2              -         -    -        -        -   \n",
       "4           k119_14       4_1              -         -    -        -        -   \n",
       "...             ...       ...            ...       ...  ...      ...      ...   \n",
       "262036  k119_301781  123146_1              -         -    -        -        -   \n",
       "262037  k119_301781  123146_1              -         -    -        -        -   \n",
       "262038  k119_301781  123146_1              -         -    -        -        -   \n",
       "262039  k119_301781  123146_2              -         -    -        -        -   \n",
       "262040  k119_301781  123146_3              -         -    -        -        -   \n",
       "\n",
       "       CompRanking_MGE_prediction mobileOG_ID Gene_Name  ... coverage MOB  \\\n",
       "0                           phage           -         -  ...        -   -   \n",
       "1                    unclassified           -         -  ...        -   -   \n",
       "2                    unclassified           -         -  ...        -   -   \n",
       "3                    unclassified           -         -  ...        -   -   \n",
       "4                         plasmid           -         -  ...        -   -   \n",
       "...                           ...         ...       ...  ...      ...  ..   \n",
       "262036               unclassified           -         -  ...        -   -   \n",
       "262037               unclassified           -         -  ...        -   -   \n",
       "262038               unclassified           -         -  ...        -   -   \n",
       "262039               unclassified           -         -  ...        -   -   \n",
       "262040               unclassified           -         -  ...        -   -   \n",
       "\n",
       "           Contig_y       ORF Virulence_factor VF_Name VF_Host_Bacteria  \\\n",
       "0            k119_7       1_1                -       -                -   \n",
       "1            k119_8       2_1                -       -                -   \n",
       "2           k119_12       3_1                -       -                -   \n",
       "3           k119_12       3_2                -       -                -   \n",
       "4           k119_14       4_1                -       -                -   \n",
       "...             ...       ...              ...     ...              ...   \n",
       "262036  k119_301781  123146_1                -       -                -   \n",
       "262037  k119_301781  123146_1                -       -                -   \n",
       "262038  k119_301781  123146_1                -       -                -   \n",
       "262039  k119_301781  123146_2                -       -                -   \n",
       "262040  k119_301781  123146_3                -       -                -   \n",
       "\n",
       "       VFcategory ESKAPE Pathogenicity  \n",
       "0               -      -             -  \n",
       "1               -      -    Pathogenic  \n",
       "2               -      -    Pathogenic  \n",
       "3               -      -    Pathogenic  \n",
       "4               -      -             -  \n",
       "...           ...    ...           ...  \n",
       "262036          -      -    Pathogenic  \n",
       "262037          -      -    Pathogenic  \n",
       "262038          -      -    Pathogenic  \n",
       "262039          -      -    Pathogenic  \n",
       "262040          -      -    Pathogenic  \n",
       "\n",
       "[262041 rows x 23 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sum=pd.merge(df_AMR_MOB,df_VF_PATH,left_on=\"ORF_ID\",right_on=\"ORF\",how=\"left\")\n",
    "df_sum\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig_x</th>\n",
       "      <th>ORF_ID</th>\n",
       "      <th>ARG_prediction</th>\n",
       "      <th>ARG_class</th>\n",
       "      <th>Database</th>\n",
       "      <th>ARG_rank</th>\n",
       "      <th>CompRanking_MGE_prediction</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Taxonomy</th>\n",
       "      <th>Major_Category</th>\n",
       "      <th>MGE_Database</th>\n",
       "      <th>MOB</th>\n",
       "      <th>Virulence_factor</th>\n",
       "      <th>VFcategory</th>\n",
       "      <th>ESKAPE</th>\n",
       "      <th>Pathogenicity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262036</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262037</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262038</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262039</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>262040</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>262041 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Contig_x    ORF_ID ARG_prediction ARG_class Database ARG_rank  \\\n",
       "0            k119_7       1_1              -         -        -        -   \n",
       "1            k119_8       2_1              -         -        -        -   \n",
       "2           k119_12       3_1              -         -        -        -   \n",
       "3           k119_12       3_2              -         -        -        -   \n",
       "4           k119_14       4_1              -         -        -        -   \n",
       "...             ...       ...            ...       ...      ...      ...   \n",
       "262036  k119_301781  123146_1              -         -        -        -   \n",
       "262037  k119_301781  123146_1              -         -        -        -   \n",
       "262038  k119_301781  123146_1              -         -        -        -   \n",
       "262039  k119_301781  123146_2              -         -        -        -   \n",
       "262040  k119_301781  123146_3              -         -        -        -   \n",
       "\n",
       "       CompRanking_MGE_prediction Gene_Name Taxonomy Major_Category  \\\n",
       "0                           phage         -        -              -   \n",
       "1                    unclassified         -        -              -   \n",
       "2                    unclassified         -        -              -   \n",
       "3                    unclassified         -        -              -   \n",
       "4                         plasmid         -        -              -   \n",
       "...                           ...       ...      ...            ...   \n",
       "262036               unclassified         -        -              -   \n",
       "262037               unclassified         -        -              -   \n",
       "262038               unclassified         -        -              -   \n",
       "262039               unclassified         -        -              -   \n",
       "262040               unclassified         -        -              -   \n",
       "\n",
       "       MGE_Database MOB Virulence_factor VFcategory ESKAPE Pathogenicity  \n",
       "0                 -   -                -          -      -             -  \n",
       "1                 -   -                -          -      -    Pathogenic  \n",
       "2                 -   -                -          -      -    Pathogenic  \n",
       "3                 -   -                -          -      -    Pathogenic  \n",
       "4                 -   -                -          -      -             -  \n",
       "...             ...  ..              ...        ...    ...           ...  \n",
       "262036            -   -                -          -      -    Pathogenic  \n",
       "262037            -   -                -          -      -    Pathogenic  \n",
       "262038            -   -                -          -      -    Pathogenic  \n",
       "262039            -   -                -          -      -    Pathogenic  \n",
       "262040            -   -                -          -      -    Pathogenic  \n",
       "\n",
       "[262041 rows x 16 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sum=df_sum.drop([\"Contig_y\",\"ORF\",\"VF_Host_Bacteria\",\"coverage\",\"SNPs\",\"mobileOG_ID\",\"VF_Name\"],axis=1,inplace=False)\n",
    "df_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>ORF_ID</th>\n",
       "      <th>ARG_prediction</th>\n",
       "      <th>ARG_class</th>\n",
       "      <th>ARG_Database</th>\n",
       "      <th>ARG_rank(recorded)</th>\n",
       "      <th>ARG_Rank(in-situ)</th>\n",
       "      <th>MGE_prediction</th>\n",
       "      <th>MGE_Name</th>\n",
       "      <th>MGE_Taxonomy</th>\n",
       "      <th>MGE_Category</th>\n",
       "      <th>MGE_Database</th>\n",
       "      <th>Mobility</th>\n",
       "      <th>Viculence_factor</th>\n",
       "      <th>VF_Category</th>\n",
       "      <th>VF_ESKAPE</th>\n",
       "      <th>Pathogenicity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240330</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240331</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240332</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>wbtL</td>\n",
       "      <td>Immune modulation</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240333</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240334</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>240335 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Contig    ORF_ID ARG_prediction ARG_class ARG_Database  \\\n",
       "0            k119_7       1_1              -         -            -   \n",
       "1            k119_8       2_1              -         -            -   \n",
       "2           k119_12       3_1              -         -            -   \n",
       "3           k119_12       3_2              -         -            -   \n",
       "4           k119_14       4_1              -         -            -   \n",
       "...             ...       ...            ...       ...          ...   \n",
       "240330  k119_301780  123145_1              -         -            -   \n",
       "240331  k119_301780  123145_2              -         -            -   \n",
       "240332  k119_301781  123146_1              -         -            -   \n",
       "240333  k119_301781  123146_2              -         -            -   \n",
       "240334  k119_301781  123146_3              -         -            -   \n",
       "\n",
       "       ARG_rank(recorded) ARG_Rank(in-situ) MGE_prediction MGE_Name  \\\n",
       "0                       -                 -          phage        -   \n",
       "1                       -                 -   unclassified        -   \n",
       "2                       -                 -   unclassified        -   \n",
       "3                       -                 -   unclassified        -   \n",
       "4                       -                 -        plasmid        -   \n",
       "...                   ...               ...            ...      ...   \n",
       "240330                  -                 -     chromosome        -   \n",
       "240331                  -                 -     chromosome        -   \n",
       "240332                  -                 -          phage        -   \n",
       "240333                  -                 -          phage        -   \n",
       "240334                  -                 -          phage        -   \n",
       "\n",
       "       MGE_Taxonomy MGE_Category MGE_Database Mobility Viculence_factor  \\\n",
       "0                 -            -            -        -                -   \n",
       "1                 -            -            -        -                -   \n",
       "2                 -            -            -        -                -   \n",
       "3                 -            -            -        -                -   \n",
       "4                 -            -            -        -                -   \n",
       "...             ...          ...          ...      ...              ...   \n",
       "240330            -            -            -        -                -   \n",
       "240331            -            -            -        -                -   \n",
       "240332            -            -            -        -             wbtL   \n",
       "240333            -            -            -        -                -   \n",
       "240334            -            -            -        -                -   \n",
       "\n",
       "              VF_Category VF_ESKAPE Pathogenicity  \n",
       "0                       -         -             -  \n",
       "1                       -         -    Pathogenic  \n",
       "2                       -         -    Pathogenic  \n",
       "3                       -         -    Pathogenic  \n",
       "4                       -         -             -  \n",
       "...                   ...       ...           ...  \n",
       "240330                  -         -             -  \n",
       "240331                  -         -             -  \n",
       "240332  Immune modulation  Negative    Pathogenic  \n",
       "240333                  -         -    Pathogenic  \n",
       "240334                  -         -    Pathogenic  \n",
       "\n",
       "[240335 rows x 17 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sum.columns=[\"Contig\",\"ORF_ID\",\"ARG_prediction\",\"ARG_class\",\"ARG_Database\",\"ARG_rank(recorded)\",\"MGE_prediction\",\"MGE_Name\",\"MGE_Taxonomy\",\"MGE_Category\",\"MGE_Database\",\"Mobility\",\"Viculence_factor\",\"VF_Category\",\"VF_ESKAPE\",\"Pathogenicity\"]\n",
    "df_sum[\"ARG_Rank(in-situ)\"]=\"-\"\n",
    "df_sum=df_sum[[\"Contig\",\"ORF_ID\",\"ARG_prediction\",\"ARG_class\",\"ARG_Database\",\"ARG_rank(recorded)\",\"ARG_Rank(in-situ)\",\"MGE_prediction\",\"MGE_Name\",\"MGE_Taxonomy\",\"MGE_Category\",\"MGE_Database\",\"Mobility\",\"Viculence_factor\",\"VF_Category\",\"VF_ESKAPE\",\"Pathogenicity\"]]\n",
    "df_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>ORF_ID</th>\n",
       "      <th>ARG_prediction</th>\n",
       "      <th>ARG_class</th>\n",
       "      <th>ARG_Database</th>\n",
       "      <th>ARG_rank(recorded)</th>\n",
       "      <th>ARG_Rank(in-situ)</th>\n",
       "      <th>MGE_prediction</th>\n",
       "      <th>MGE_Name</th>\n",
       "      <th>MGE_Taxonomy</th>\n",
       "      <th>MGE_Category</th>\n",
       "      <th>MGE_Database</th>\n",
       "      <th>Mobility</th>\n",
       "      <th>Viculence_factor</th>\n",
       "      <th>VF_Category</th>\n",
       "      <th>VF_ESKAPE</th>\n",
       "      <th>Pathogenicity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>k119_7</td>\n",
       "      <td>1_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>k119_8</td>\n",
       "      <td>2_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>k119_12</td>\n",
       "      <td>3_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>unclassified</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>k119_14</td>\n",
       "      <td>4_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>plasmid</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240330</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240331</th>\n",
       "      <td>k119_301780</td>\n",
       "      <td>123145_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240332</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>wbtL</td>\n",
       "      <td>Immune modulation</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240333</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240334</th>\n",
       "      <td>k119_301781</td>\n",
       "      <td>123146_3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>IV</td>\n",
       "      <td>phage</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Pathogenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>240335 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Contig    ORF_ID ARG_prediction ARG_class ARG_Database  \\\n",
       "0            k119_7       1_1              -         -            -   \n",
       "1            k119_8       2_1              -         -            -   \n",
       "2           k119_12       3_1              -         -            -   \n",
       "3           k119_12       3_2              -         -            -   \n",
       "4           k119_14       4_1              -         -            -   \n",
       "...             ...       ...            ...       ...          ...   \n",
       "240330  k119_301780  123145_1              -         -            -   \n",
       "240331  k119_301780  123145_2              -         -            -   \n",
       "240332  k119_301781  123146_1              -         -            -   \n",
       "240333  k119_301781  123146_2              -         -            -   \n",
       "240334  k119_301781  123146_3              -         -            -   \n",
       "\n",
       "       ARG_rank(recorded) ARG_Rank(in-situ) MGE_prediction MGE_Name  \\\n",
       "0                       -                IV          phage        -   \n",
       "1                       -                IV   unclassified        -   \n",
       "2                       -                IV   unclassified        -   \n",
       "3                       -                IV   unclassified        -   \n",
       "4                       -                IV        plasmid        -   \n",
       "...                   ...               ...            ...      ...   \n",
       "240330                  -                IV     chromosome        -   \n",
       "240331                  -                IV     chromosome        -   \n",
       "240332                  -                IV          phage        -   \n",
       "240333                  -                IV          phage        -   \n",
       "240334                  -                IV          phage        -   \n",
       "\n",
       "       MGE_Taxonomy MGE_Category MGE_Database Mobility Viculence_factor  \\\n",
       "0                 -            -            -        -                -   \n",
       "1                 -            -            -        -                -   \n",
       "2                 -            -            -        -                -   \n",
       "3                 -            -            -        -                -   \n",
       "4                 -            -            -        -                -   \n",
       "...             ...          ...          ...      ...              ...   \n",
       "240330            -            -            -        -                -   \n",
       "240331            -            -            -        -                -   \n",
       "240332            -            -            -        -             wbtL   \n",
       "240333            -            -            -        -                -   \n",
       "240334            -            -            -        -                -   \n",
       "\n",
       "              VF_Category VF_ESKAPE Pathogenicity  \n",
       "0                       -         -             -  \n",
       "1                       -         -    Pathogenic  \n",
       "2                       -         -    Pathogenic  \n",
       "3                       -         -    Pathogenic  \n",
       "4                       -         -             -  \n",
       "...                   ...       ...           ...  \n",
       "240330                  -         -             -  \n",
       "240331                  -         -             -  \n",
       "240332  Immune modulation  Negative    Pathogenic  \n",
       "240333                  -         -    Pathogenic  \n",
       "240334                  -         -    Pathogenic  \n",
       "\n",
       "[240335 rows x 17 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cpr_rank=[]\n",
    "for i, name in df_sum.iterrows():\n",
    "    if name[\"ARG_prediction\"] == \"-\":\n",
    "        cpr_rank.append(\"IV\")\n",
    "    elif name[\"ARG_prediction\"]!= \"-\" and name[\"MGE_Name\"] == \"-\":\n",
    "        cpr_rank.append(\"III\")\n",
    "    elif name[\"ARG_prediction\"]!= \"-\" and name[\"MGE_Name\"] != \"-\" and name[\"Pathogenicity\"] == \"-\":\n",
    "        cpr_rank.append(\"II\")\n",
    "    elif name[\"ARG_prediction\"]!= \"-\" and name[\"MGE_Name\"] != \"-\" and name[\"Pathogenicity\"] != \"-\":\n",
    "        cpr_rank.append(\"I\")\n",
    "    else:\n",
    "        cpr_rank.append(\"-\")\n",
    "df_sum[\"ARG_Rank(in-situ)\"]=cpr_rank\n",
    "\n",
    "df_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "38978"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=df_sum[df_sum.Pathogenicity == \"Pathogenic\"]\n",
    "b=a.Contig.unique()\n",
    "len(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:18) \n[GCC 10.3.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "96a28c887955343dfdcfe0648667fa1d323b86bd8c9a5b8cfe9af24fde7c7749"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
