{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing SARG outputfile for ARG ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Precessing dmnd file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_303897/45944870.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_sarg_len[\"coverage\"][index]=df_sarg_len[\"alignLen\"][index]/df_sarg_len[2][index]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig_ID</th>\n",
       "      <th>query</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Contig_ID, query]\n",
       "Index: []"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#coveragy cal\n",
    "input_sarg=\"/lomi_home/gaoyang/software/CompRanking/test/CompRanking/CompRanking_intermediate/AMR/ARGranking/ERR1191817.contigs_5M_contigs_SARG_Protein_diamond.txt\"\n",
    "input_sarg=\"/lomi_home/gaoyang/software/CompRanking/benchmarking/MGE_benchmarking/plasmid_MOB/non_mob_plas/nonmobplas/CompRanking_intermediate/AMR/ARGranking/nonmob_plas_5M_contigs_SARG_Protein_diamond.txt\"\n",
    "df_sarg=pd.read_csv(input_sarg, sep=\"\\t\",header=None)\n",
    "df_sarg.columns = ['id', 'sub_id', 'identity', 'alignLen', 'mismat', 'gapOpens', 'qStart', 'qEnd', 'sStart', 'sEnd', 'eval', 'bit']\n",
    "\n",
    "#open df_sarg\n",
    "df_SARG_len=pd.read_csv(\"/lomi_home/gaoyang/software/CompRanking/databases/SARG/SARG.db.fasta.length\", sep=\"\\t\", header=None )\n",
    "df_SARG_len=df_SARG_len[[1,2]]\n",
    "\n",
    "#merge len and df_sarg\n",
    "df_sarg_len=pd.merge(df_sarg,df_SARG_len,left_on=\"sub_id\",right_on=1,how=\"left\")\n",
    "df_sarg_len[\"coverage\"]=\"-\"\n",
    "\n",
    "#cover cal\n",
    "for index, name in df_sarg_len.iterrows():\n",
    "    df_sarg_len[\"coverage\"][index]=df_sarg_len[\"alignLen\"][index]/df_sarg_len[2][index]\n",
    "\n",
    "# filter out contigs identity under 80\n",
    "df_sarg_iden95 = df_sarg_len[df_sarg_len.identity > 80]\n",
    "# filter out contigs alignment length under 0.75\n",
    "df_sarg_alen = df_sarg_iden95[df_sarg_iden95.coverage>0.75]\n",
    "df_sarg_alen\n",
    "df_sarg_tmp=df_sarg_alen.iloc[:,[0,1]]\n",
    "df_sarg_tmp.columns=[\"Contig_ID\", \"query\"]\n",
    "df_sarg_tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Reference</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0910185A</td>\n",
       "      <td>aminoglycoside__aph(3)-IV</td>\n",
       "      <td>aminoglycoside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1112175A</td>\n",
       "      <td>macrolide-lincosamide-streptogramin__ermE</td>\n",
       "      <td>macrolide-lincosamide-streptogramin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1405331A</td>\n",
       "      <td>aminoglycoside__streptomycin resistance protein</td>\n",
       "      <td>aminoglycoside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1411197A</td>\n",
       "      <td>beta-lactam__class D beta-lactamase</td>\n",
       "      <td>beta-lactam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1616230A</td>\n",
       "      <td>beta-lactam__class A beta-lactamase</td>\n",
       "      <td>beta-lactam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4045</th>\n",
       "      <td>ZP_04668102</td>\n",
       "      <td>tetracycline__tetO</td>\n",
       "      <td>tetracycline</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4046</th>\n",
       "      <td>ZP_04677690</td>\n",
       "      <td>multidrug__norA</td>\n",
       "      <td>multidrug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4047</th>\n",
       "      <td>ZP_04677702</td>\n",
       "      <td>bacitracin__bacA</td>\n",
       "      <td>bacitracin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4048</th>\n",
       "      <td>ZP_04679156</td>\n",
       "      <td>multidrug__mepA</td>\n",
       "      <td>multidrug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049</th>\n",
       "      <td>WP_049589868.1_1</td>\n",
       "      <td>colistin__mcr1</td>\n",
       "      <td>colistin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4050 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Reference                                         Genotype  \\\n",
       "0             0910185A                        aminoglycoside__aph(3)-IV   \n",
       "1             1112175A        macrolide-lincosamide-streptogramin__ermE   \n",
       "2             1405331A  aminoglycoside__streptomycin resistance protein   \n",
       "3             1411197A              beta-lactam__class D beta-lactamase   \n",
       "4             1616230A              beta-lactam__class A beta-lactamase   \n",
       "...                ...                                              ...   \n",
       "4045       ZP_04668102                               tetracycline__tetO   \n",
       "4046       ZP_04677690                                  multidrug__norA   \n",
       "4047       ZP_04677702                                 bacitracin__bacA   \n",
       "4048       ZP_04679156                                  multidrug__mepA   \n",
       "4049  WP_049589868.1_1                                   colistin__mcr1   \n",
       "\n",
       "                                Phenotype  \n",
       "0                          aminoglycoside  \n",
       "1     macrolide-lincosamide-streptogramin  \n",
       "2                          aminoglycoside  \n",
       "3                             beta-lactam  \n",
       "4                             beta-lactam  \n",
       "...                                   ...  \n",
       "4045                         tetracycline  \n",
       "4046                            multidrug  \n",
       "4047                           bacitracin  \n",
       "4048                            multidrug  \n",
       "4049                             colistin  \n",
       "\n",
       "[4050 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_sarg_structure=\"/lomi_home/gaoyang/software/CompRanking/databases/SARG/SARG.structure.txt\"\n",
    "df_sarg_structure=pd.read_csv(input_sarg_structure, sep=\"\\t\", header=0)\n",
    "df_sarg_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig_ID</th>\n",
       "      <th>query</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Contig_ID, query, Genotype, Phenotype]\n",
       "Index: []"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sarg_type=pd.merge(df_sarg_tmp,df_sarg_structure,left_on=\"query\",right_on=\"Reference\",how=\"left\")\n",
    "df_sarg_type1=df_sarg_type.drop(['Reference'], axis=1, inplace=False)\n",
    "df_sarg_type1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "true\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig_ID</th>\n",
       "      <th>query</th>\n",
       "      <th>class</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>SARG_Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Contig_ID, query, class, Phenotype, SARG_Rank]\n",
       "Index: []"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_argrank1.columns=[\"Contig_ID\",\"query\",\"class\",\"Phenotype\",\"SARG_Rank\"]\n",
    "if len(df_sarg_type1[\"Contig_ID\"]) == 0:\n",
    "    print(\"true\")\n",
    "    df_sarg_type1[\"Phenotype\"]=[]\n",
    "    df_sarg_type1[\"SARG_Rank\"]=[]\n",
    "    df_sarg_type1.columns=[\"Contig_ID\",\"query\",\"class\",\"Phenotype\",\"SARG_Rank\"]\n",
    "else:\n",
    "    pass\n",
    "\n",
    "df_sarg_type1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda/lib/python3.8/site-packages/pandas/core/indexes/base.py:3621\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3620\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 3621\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[1;32m   3622\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[0;32m~/miniconda/lib/python3.8/site-packages/pandas/_libs/index.pyx:136\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m~/miniconda/lib/python3.8/site-packages/pandas/_libs/index.pyx:163\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5198\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5206\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 1",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [5], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m#standardizing format\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m df_sarg_type1[\u001b[39m\"\u001b[39m\u001b[39mclass\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m=\u001b[39mdf_sarg_type1[\u001b[39m\"\u001b[39;49m\u001b[39mGenotype\u001b[39;49m\u001b[39m\"\u001b[39;49m]\u001b[39m.\u001b[39;49mstr\u001b[39m.\u001b[39;49msplit(\u001b[39m\"\u001b[39;49m\u001b[39m__\u001b[39;49m\u001b[39m\"\u001b[39;49m,expand\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m)[\u001b[39m1\u001b[39;49m]\n\u001b[1;32m      3\u001b[0m df_sarg_type2\u001b[39m=\u001b[39mdf_sarg_type1\u001b[39m.\u001b[39mdrop([\u001b[39m'\u001b[39m\u001b[39mGenotype\u001b[39m\u001b[39m'\u001b[39m], axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m, inplace\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n\u001b[1;32m      4\u001b[0m df_sarg_type2\u001b[39m=\u001b[39mdf_sarg_type2[[\u001b[39m\"\u001b[39m\u001b[39mContig_ID\u001b[39m\u001b[39m\"\u001b[39m,\u001b[39m\"\u001b[39m\u001b[39mquery\u001b[39m\u001b[39m\"\u001b[39m,\u001b[39m\"\u001b[39m\u001b[39mclass\u001b[39m\u001b[39m\"\u001b[39m,\u001b[39m\"\u001b[39m\u001b[39mPhenotype\u001b[39m\u001b[39m\"\u001b[39m]]\n",
      "File \u001b[0;32m~/miniconda/lib/python3.8/site-packages/pandas/core/frame.py:3505\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3503\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m   3504\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3505\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[1;32m   3506\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3507\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda/lib/python3.8/site-packages/pandas/core/indexes/base.py:3623\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3621\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3622\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[0;32m-> 3623\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[1;32m   3624\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[1;32m   3625\u001b[0m     \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3626\u001b[0m     \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3627\u001b[0m     \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3628\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 1"
     ]
    }
   ],
   "source": [
    "#standardizing format\n",
    "df_sarg_type1[\"class\"]=df_sarg_type1[\"Genotype\"].str.split(\"__\",expand=True)[1]\n",
    "df_sarg_type2=df_sarg_type1.drop(['Genotype'], axis=1, inplace=False)\n",
    "df_sarg_type2=df_sarg_type2[[\"Contig_ID\",\"query\",\"class\",\"Phenotype\"]]\n",
    "df_sarg_type2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Processing arg ranking index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARG</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>Enrichment</th>\n",
       "      <th>Mobility</th>\n",
       "      <th>ESKAPE_pathogen</th>\n",
       "      <th>Plasmid</th>\n",
       "      <th>Integron</th>\n",
       "      <th>Host_range</th>\n",
       "      <th>Human_associated_occ</th>\n",
       "      <th>Human_associated_abun</th>\n",
       "      <th>Enrichment_value</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAB20442</td>\n",
       "      <td>aminoglycoside__aac(3)-II</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>False</td>\n",
       "      <td>class</td>\n",
       "      <td>0.0470</td>\n",
       "      <td>1.00E-03</td>\n",
       "      <td>Human_only</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YP_001966247</td>\n",
       "      <td>aminoglycoside__aac(3)-II</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>False</td>\n",
       "      <td>class</td>\n",
       "      <td>0.0840</td>\n",
       "      <td>4.40E-03</td>\n",
       "      <td>Human_only</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_009838.1.5616796.p01</td>\n",
       "      <td>aminoglycoside__aac(3)-VI</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>False</td>\n",
       "      <td>family</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>4.20E-03</td>\n",
       "      <td>322.544845</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABG77519</td>\n",
       "      <td>aminoglycoside__aac(6')-I</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>True</td>\n",
       "      <td>phylum</td>\n",
       "      <td>0.0690</td>\n",
       "      <td>4.10E-03</td>\n",
       "      <td>416.195466</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF231133.1.gene2.p01</td>\n",
       "      <td>aminoglycoside__aac(6')-I</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>TRUE</td>\n",
       "      <td>False</td>\n",
       "      <td>phylum</td>\n",
       "      <td>0.0160</td>\n",
       "      <td>3.90E-04</td>\n",
       "      <td>Human_only</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4045</th>\n",
       "      <td>DQ018711.1.gene7.p01</td>\n",
       "      <td>vancomycin__vanY</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4700</td>\n",
       "      <td>4.40E-03</td>\n",
       "      <td>21.69590399</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4046</th>\n",
       "      <td>FJ349556.1.gene4.p01</td>\n",
       "      <td>vancomycin__vanY</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4900</td>\n",
       "      <td>5.40E-03</td>\n",
       "      <td>7.498609207</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4047</th>\n",
       "      <td>AF155139.2.gene5.p01</td>\n",
       "      <td>vancomycin__vanZ</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1400</td>\n",
       "      <td>2.70E-03</td>\n",
       "      <td>1.699360876</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4048</th>\n",
       "      <td>CAB61225</td>\n",
       "      <td>vancomycin__vanZ</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>4.50E-05</td>\n",
       "      <td>0.001356539</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049</th>\n",
       "      <td>DQ018711.1.gene8.p01</td>\n",
       "      <td>vancomycin__vanZ</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4050 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ARG                   Genotype       Phenotype  \\\n",
       "0                    AAB20442  aminoglycoside__aac(3)-II  aminoglycoside   \n",
       "1                YP_001966247  aminoglycoside__aac(3)-II  aminoglycoside   \n",
       "2     NC_009838.1.5616796.p01  aminoglycoside__aac(3)-VI  aminoglycoside   \n",
       "3                    ABG77519  aminoglycoside__aac(6')-I  aminoglycoside   \n",
       "4        AF231133.1.gene2.p01  aminoglycoside__aac(6')-I  aminoglycoside   \n",
       "...                       ...                        ...             ...   \n",
       "4045     DQ018711.1.gene7.p01           vancomycin__vanY      vancomycin   \n",
       "4046     FJ349556.1.gene4.p01           vancomycin__vanY      vancomycin   \n",
       "4047     AF155139.2.gene5.p01           vancomycin__vanZ      vancomycin   \n",
       "4048                 CAB61225           vancomycin__vanZ      vancomycin   \n",
       "4049     DQ018711.1.gene8.p01           vancomycin__vanZ      vancomycin   \n",
       "\n",
       "     Enrichment Mobility ESKAPE_pathogen Plasmid  Integron Host_range  \\\n",
       "0          TRUE     TRUE            TRUE    TRUE     False      class   \n",
       "1          TRUE     TRUE            TRUE    TRUE     False      class   \n",
       "2          TRUE     TRUE            TRUE    TRUE     False     family   \n",
       "3          TRUE     TRUE            TRUE    TRUE      True     phylum   \n",
       "4          TRUE     TRUE            TRUE    TRUE     False     phylum   \n",
       "...         ...      ...             ...     ...       ...        ...   \n",
       "4045       None     None            None    None     False        NaN   \n",
       "4046       None     None            None    None     False        NaN   \n",
       "4047       None     None            None    None     False        NaN   \n",
       "4048       None     None            None    None     False        NaN   \n",
       "4049       None     None            None    None     False        NaN   \n",
       "\n",
       "      Human_associated_occ Human_associated_abun Enrichment_value        Rank  \n",
       "0                   0.0470              1.00E-03       Human_only           I  \n",
       "1                   0.0840              4.40E-03       Human_only           I  \n",
       "2                   0.0041              4.20E-03       322.544845           I  \n",
       "3                   0.0690              4.10E-03       416.195466           I  \n",
       "4                   0.0160              3.90E-04       Human_only           I  \n",
       "...                    ...                   ...              ...         ...  \n",
       "4045                0.4700              4.40E-03      21.69590399  Unassessed  \n",
       "4046                0.4900              5.40E-03      7.498609207  Unassessed  \n",
       "4047                0.1400              2.70E-03      1.699360876  Unassessed  \n",
       "4048                0.0020              4.50E-05      0.001356539  Unassessed  \n",
       "4049                0.0000                  None             None  Unassessed  \n",
       "\n",
       "[4050 rows x 13 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#open arg rank index file\n",
    "input_argrank=\"/lomi_home/gaoyang/software/CompRanking/databases/SARG/ARG_rank.txt\"\n",
    "df_argrank=pd.read_csv(input_argrank, sep=\"\\t\", header=0)\n",
    "df_argrank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARG</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAB20442</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YP_001966247</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_009838.1.5616796.p01</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABG77519</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AF231133.1.gene2.p01</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4045</th>\n",
       "      <td>DQ018711.1.gene7.p01</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4046</th>\n",
       "      <td>FJ349556.1.gene4.p01</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4047</th>\n",
       "      <td>AF155139.2.gene5.p01</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4048</th>\n",
       "      <td>CAB61225</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4049</th>\n",
       "      <td>DQ018711.1.gene8.p01</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4050 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ARG        Rank\n",
       "0                    AAB20442           I\n",
       "1                YP_001966247           I\n",
       "2     NC_009838.1.5616796.p01           I\n",
       "3                    ABG77519           I\n",
       "4        AF231133.1.gene2.p01           I\n",
       "...                       ...         ...\n",
       "4045     DQ018711.1.gene7.p01  Unassessed\n",
       "4046     FJ349556.1.gene4.p01  Unassessed\n",
       "4047     AF155139.2.gene5.p01  Unassessed\n",
       "4048                 CAB61225  Unassessed\n",
       "4049     DQ018711.1.gene8.p01  Unassessed\n",
       "\n",
       "[4050 rows x 2 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_arg1=df_argrank[[\"ARG\",\"Rank\"]]\n",
    "df_arg1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig_ID</th>\n",
       "      <th>query</th>\n",
       "      <th>class</th>\n",
       "      <th>Phenotype</th>\n",
       "      <th>SARG_Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>240_1</td>\n",
       "      <td>ABB89122</td>\n",
       "      <td>dfrA5</td>\n",
       "      <td>trimethoprim</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>240_2</td>\n",
       "      <td>AB104852.1.gene2.p01</td>\n",
       "      <td>IMP-1</td>\n",
       "      <td>beta-lactam</td>\n",
       "      <td>Unassessed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1157_2</td>\n",
       "      <td>ACP19239</td>\n",
       "      <td>vanZ</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1157_3</td>\n",
       "      <td>AAA65958</td>\n",
       "      <td>vanY</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1157_4</td>\n",
       "      <td>ACC93634</td>\n",
       "      <td>vanX</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>117583_3</td>\n",
       "      <td>AAF72359</td>\n",
       "      <td>vanS</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>117583_4</td>\n",
       "      <td>KF823969.1.gene1.p01</td>\n",
       "      <td>vanR</td>\n",
       "      <td>vancomycin</td>\n",
       "      <td>III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>121135_4</td>\n",
       "      <td>YP_112402</td>\n",
       "      <td>aph(3)-I</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>IV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>121356_1</td>\n",
       "      <td>CAG34229</td>\n",
       "      <td>ant(2)-I</td>\n",
       "      <td>aminoglycoside</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>121371_3</td>\n",
       "      <td>L06798.gene.p01</td>\n",
       "      <td>tetD</td>\n",
       "      <td>tetracycline</td>\n",
       "      <td>IV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Contig_ID                 query     class       Phenotype   SARG_Rank\n",
       "0       240_1              ABB89122     dfrA5    trimethoprim           I\n",
       "1       240_2  AB104852.1.gene2.p01     IMP-1     beta-lactam  Unassessed\n",
       "2      1157_2              ACP19239      vanZ      vancomycin          IV\n",
       "3      1157_3              AAA65958      vanY      vancomycin           I\n",
       "4      1157_4              ACC93634      vanX      vancomycin          IV\n",
       "..        ...                   ...       ...             ...         ...\n",
       "106  117583_3              AAF72359      vanS      vancomycin         III\n",
       "107  117583_4  KF823969.1.gene1.p01      vanR      vancomycin         III\n",
       "108  121135_4             YP_112402  aph(3)-I  aminoglycoside          IV\n",
       "109  121356_1              CAG34229  ant(2)-I  aminoglycoside           I\n",
       "110  121371_3       L06798.gene.p01      tetD    tetracycline          IV\n",
       "\n",
       "[111 rows x 5 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_argrank2=pd.merge(df_sarg_type2,df_arg1,left_on=\"query\",right_on=\"ARG\",how=\"inner\")\n",
    "df_argrank2=df_argrank2.drop([\"ARG\"],axis=1,inplace=False)\n",
    "df_argrank2.columns=[\"Contig_ID\",\"query\",\"class\",\"Phenotype\",\"SARG_Rank\"]\n",
    "df_argrank2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_argrank2.to_csv(\"./test_SARGrank_Protein90_Result_tmp.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:18) \n[GCC 10.3.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "96a28c887955343dfdcfe0648667fa1d323b86bd8c9a5b8cfe9af24fde7c7749"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
